modules/programs/go.nix

```diff
@@ -0,0 +1,67 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.programs.go;
+
+in
+
+{
+  meta.maintainers = [ maintainers.rvolosatovs ];
+
+  options = {
+    programs.go = {
+      enable = mkEnableOption "Go";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.go;
+        defaultText = "pkgs.go";
+        description = "The Go package to use.";
+      };
+
+      packages = mkOption {
+        type = with types; attrsOf path;
+        example = literalExample ''
+          {
+            "golang.org/x/time/rate" = builtins.fetchGit "https://go.googlesource.com/text";
+          }
+        '';
+        description = "Packages to add to GOPATH.";
+      };
+
+      goPath = mkOption {
+        type = with types; nullOr str;
+        default = "go";
+        example = "";
+        description = "GOPATH relative to HOME";
+      };
+
+      goBin = mkOption {
+        type = with types; nullOr str;
+        default = "";
+        example = ".local/bin.go";
+        description = "GOBIN relative to HOME";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable (mkMerge [
+    {
+      home.packages = [ cfg.package ];
+
+      home.file = let
+        goPath = if cfg ? goPath then cfg.goPath else "go";
+      in
+      mapAttrs' (name: source: nameValuePair "${goPath}/src/${name}" { inherit source; }) cfg.packages;
+    }
+    (mkIf (cfg ? goPath) {
```

Comment:
Is the intent to not set the `GOPATH` and `GOBIN` variables if the user has left these unset? Right now I believe these will always be set because `cfg ? goPath` and `cfg ? goBin` is always true.

Maybe these options should have `default = null` and this then becomes, e.g., `mkIf (cfg.goPath != null) { … }`?

Reply:
On the other hand, maybe always set the GOPATH and GOBIN variables? I.e., have `type = types.str` for `programs.go.goPath` and `programs.go.goBin`. Then no need to worry about the `mkIf` at all 🙂

Reply:
If `$GOPATH` is not set, `go` uses `$HOME/go` as `GOPATH`.
If `$GOBIN` is not set, `go` uses `$GOPATH/bin` as `GOBIN`.

This is the behavior I was trying to replicate.
I thought `cfg ? goPath` would evaluate to `false` if `goPath` is not set in `cfg`, and `true`, if it is - is that not correct?

Reply:
No, the `goPath` attribute is set in `cfg` because it has a default value in the option.

For your intended semantics I would suggest simply having

```nix
{
  options.programs.go = {

    # …

    goPath = mkOption {
      type = types.str;
      default = "go";
      example = "";
      description = "GOPATH relative to HOME";
    };

    goBin = mkOption {
      type = types.str;
      default = cfg.goPath + "/bin";
      defaultText = "go/bin";
      example = ".local/bin.go";
      description = "GOBIN relative to HOME";
    };
  };

  config = mkIf cfg.enable {
    home.packages = [ cfg.package ];

    home.file =
      let
        mkSrc = n: v: {
          target = "${cfg.goPath}/src/${n}";
          source = v;
        };
      in
        mapAttrsToList mkSrc cfg.packages;

    home.sessionVariables = {
      GOPATH = "${config.home.homeDirectory}/${cfg.goPath}";
      GOBIN = "${config.home.homeDirectory}/${cfg.goBin}";
    };
  };
}
```